sdk/auth/managed_identity_authorizer.go

```diff
@@ -191,9 +192,77 @@ func azureMetadata(ctx context.Context, url string) (body []byte, err error) {
 		return
 	}
 	defer resp.Body.Close()
-	if c := resp.StatusCode; c < 200 || c > 299 {
+	if statusCode = resp.StatusCode; statusCode < 200 || statusCode > 299 {
 		err = fmt.Errorf("received HTTP status %d with body: %s", resp.StatusCode, body)
 		return
 	}
 	return
 }
+
+func retryForIMDS(ctx context.Context, url string) (body []byte, err error) {
+	// see https://docs.microsoft.com/en-us/azure/active-directory/managed-service-identity/how-to-use-vm-token#retry-guidance
+	retries := []int{
+		http.StatusRequestTimeout,      // 408
+		http.StatusTooManyRequests,     // 429
+		http.StatusInternalServerError, // 500
+		http.StatusBadGateway,          // 502
+		http.StatusServiceUnavailable,  // 503
+		http.StatusGatewayTimeout,      // 504
+	}
+
+	// extra retry status codes specific to IMDS
+	retries = append(retries,
+		http.StatusNotFound,
+		http.StatusGone,
+		// all remaining 5xx
+		http.StatusNotImplemented,
+		http.StatusHTTPVersionNotSupported,
+		http.StatusVariantAlsoNegotiates,
+		http.StatusInsufficientStorage,
+		http.StatusLoopDetected,
+		http.StatusNotExtended,
+		http.StatusNetworkAuthenticationRequired)
```

Comment:
we can remove this function - but (almost) all of these would be terminal failure modes, so shouldn't be something we should retry on